src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,593 @@
+        }
+    }
+    else // in the case we cannot use local cache, read from the original readbuffer directly
+    {
+        if (original_readbuffer == nullptr)
+            throw Exception("original readbuffer should not be null", ErrorCodes::LOGICAL_ERROR);
+        auto status = original_readbuffer->next();
+        // we don't need to worry about the memory buffer allocated in RemoteReadBuffer, since it is owned by
+        // BufferWithOwnMemory, BufferWithOwnMemory would release it.
+        //LOG_TRACE(&Poco::Logger::get("RemoteReadBuffer"), "from original rb {} {}", original_readbuffer->buffer().size(), original_readbuffer->offset());
+        if (status)
+            BufferBase::set(original_readbuffer->buffer().begin(), original_readbuffer->buffer().size(), original_readbuffer->offset());
+        return status;
+    }
+    return true;
+}
+
+off_t RemoteReadBuffer::seek(off_t offset, int whence)
+{
+    off_t pos_in_file = file_reader->getOffset();
+    off_t new_pos;
+    if (whence == SEEK_SET)
+        new_pos = offset;
+    else if (whence == SEEK_CUR)
+        new_pos = pos_in_file - (working_buffer.end() - pos) + offset;
+    else
+        throw Exception("expects SEEK_SET or SEEK_CUR as whence", ErrorCodes::BAD_ARGUMENTS);
+
+    /// Position is unchanged.
+    if (new_pos + (working_buffer.end() - pos) == pos_in_file)
+        return new_pos;
+
+    if (new_pos <= pos_in_file && new_pos >= pos_in_file - static_cast<off_t>(working_buffer.size()))
+    {
+        /// Position is still inside buffer.
+        pos = working_buffer.begin() + (new_pos - (pos_in_file - working_buffer.size()));
+        return new_pos;
+    }
+
+    pos = working_buffer.end();
+    auto ret_off = file_reader->seek(new_pos);
+    if (ret_off == -1)
+        throw Exception(
+            "seek file failed. " + std::to_string(pos_in_file) + "->" + std::to_string(new_pos) + "@" + std::to_string(file_reader->size())
+                + "," + std::to_string(whence) + "," + file_reader->getPath(),
+            ErrorCodes::BAD_ARGUMENTS);
+    return ret_off;
+}
+
+off_t RemoteReadBuffer::getPosition()
+{
+    return file_reader->getOffset() - available();
+}
+
+RemoteReadBufferCache::RemoteReadBufferCache() = default;
+
+RemoteReadBufferCache::~RemoteReadBufferCache() = default;
+
+RemoteReadBufferCache & RemoteReadBufferCache::instance()
+{
+    static RemoteReadBufferCache instance;
+    return instance;
+}
+
+void RemoteReadBufferCache::recover_cached_files_meta(
+        const std::filesystem::path &current_path,
+        size_t current_depth,
+        size_t max_depth,
+        std::function<void(RemoteCacheController *)> const & finish_callback)
+{
+    if (current_depth >= max_depth)
+    {
+        for (auto const & dir : std::filesystem::directory_iterator{current_path})
+        {
+            std::string path = dir.path();
+            if (caches.count(path))
+            {
+                LOG_ERROR(log, "duplicated file:{}", path);
```

Comment:
I still do not see why, as you say, we can't include `schema` and `cluster` in the hashed part of the path. I think it must be  included. 
Yes, may be it makes sense to have additional prefix before hashed part (apart from caches directory from config), but it seems that one directory level is enough and it can include ch table name and db name, but this is more for convenience of user, so not necessary and I am not sure we really need it, though will be ok.
Is there a reason to have these subpaths in `caches` map? To me it seems that only full paths are needed, though I haven't read the code fully.
Also as I see this method is called only from `initOnce`, which is called only once at server startup.
So I do not see any reason why we can have duplicated file here if we a traversing local filesystem and `caches` is empty before we call it from `initOnce`.

Reply:
1. schema an cluster are included to generate the hase path. the local path is two levels now

Reply:
what is `hase path`?

Reply:
ah, its a typo, you meant hash path.

Reply:
But you did not answer why do you need these additional two levels of directories with schema and cluster, if we can just include them in the hashed path of the path.

Reply:
In some cases , we may want to clear all the files of  one of the clusters without affecting others, we also can quickly to check the disk usage of every cluster, if we use the schema and cluster  as part of the path.  But this is not really necessary

The two level hash code path is to avoid too many files in one directory.

As your adivse, we can make it into one, it's OK

Reply:
> In some cases , we may want to clear all the files of one of the clusters without affecting others, we also can quickly to check the disk usage of every cluster, if we use the schema and cluster as part of the path. But this is not really necessary

Ok, lets leave this as is for now, may be will return to this a little later.

> The two level hash code path is to avoid too many files in one directory.
As your adivse, we can make it into one, it's OK

No, I didn't mean to remove two level hashed path, this is good.

Also another point. `initOnce` is called at server startup and there are 2 problems:
1. `initOnce` can throw exception and server will not startup.
2. it should be actually executed in the background, not immediately at server startup.

Also please start adding integration tests (`tests/integration/`), because each feature must have tests.

Reply:
```
1. initOnce can throw exception and server will not startup.
```
I tend to let it crash at the startup when something is going wrong. If let exception be silent is backgroup maybe a good choice.

Reply:
```
it should be actually executed in the background, not immediately at server startup.
```

we can take a try

Reply:
> I tend to let it crash at the startup when something is going wrong. If let exception be silent in backgroup maybe not a good choice.

Then you can make it under a setting (like lazy_intialization) with default as true, because we prefer never to allow such behaviour.

Reply:
> 
> Also please start adding integration tests (`tests/integration/`), because each feature must have tests.
> 

I cannot running the integration tests locally, Do I missing some environment setting?

```
$ docker run -it --rm clickhouse/integration-tests-runner:latest bash
/usr/local/bin/dockerd-entrypoint.sh: line 15: /ClickHouse/tests/integration/dockerd.log: No such file or directory
Can't start docker daemon, timeout exceeded.
```

Reply:
```
cd tests/integration/ 
export CLICKHOUSE_TESTS_SERVER_BIN_PATH=$HOME/workspace/CH/build/programs/clickhouse 
./runner "-ss test_storage_hdfs -v -vv" -- to run all tests in a file
./runner "-ss test_storage_hdfs -k test_seekable_formats -v -vv" -- to run one particular test
```

Reply:
https://github.com/ClickHouse/ClickHouse/blob/master/tests/integration/README.md

You can use `runner` or `pytest` to run tests. I recommend `runner` because you won't need to install all those packages (listed in the beginning of the page by link above) locally.

Reply:
I have wrote the tests. But then I found that I cannot start a container from the image `clickhouse/integration-tests-runner`, and the running environment is setup following the instructions in the README.md .

Reply:
>  But then I found that I cannot start a container from the image clickhouse/integration-tests-runner

what error do you get?

Reply:
```
Can't start docker daemon, timeout exceeded.
```

Reply:
`docker run hello-world` works?

Reply:
Yes, it works

Reply:
> `docker run hello-world` works? did you add a new docker-compose for hive-metastore?

Yes. But this is not relate to this part. It also fail on the example given in the readme.
```
$./runner --binary $HOME/ClickHouse/programs/clickhouse  --bridge-binary $HOME/ClickHouse/programs/clickhouse-odbc-bridge --base-configs-dir $HOME/ClickHouse/programs/server/ 'test_odbc_interaction -ss
```

Reply:
did you rebuild the image after adding a new docker-compose.yml?
if yes, is is the same after `docker pull clickhouse/clickhouse-integration-tests-runner`?

Reply:
> did you rebuild the image after adding a new docker-compose.yml? if yes, is the error still present after `docker pull clickhouse/clickhouse-integration-tests-runner`?

Do you mean to rebuild `clickhouse/clickhouse-integration-tests-runner` ? We just use `docker pull clickhouse/clickhouse-integration-tests-runner` to get it.

Reply:
Yes, but did you add a new docker-compose.yml for hive metastore? If yes, then need to rebuild the image from docker/test/integration/runner/Dockerfile.

But this is not related to error.
When I do `docker run -it --rm clickhouse/integration-tests-runner bash` I also get the same error as yours, but I can start tests with `runner` or `pytest` successfully. (Docker version 20.10.5, docker-compose version 1.28.2).

Reply:
This is weird. I run it in the source dir `ClickHouse/tests/integration` by the following cmd
```
./runner --binary $SRC_ROOT/build/programs/clickhouse  --odbc-bridge-binary $SRC_ROOT/build/programs/clickhouse-odbc-bridge 'test_odbc_interaction -ss'
```
and the ouput is 
```
2021-11-17 20:41:01,800 [ 1392747 ] INFO : ClickHouse root is not set. Will use $SRC_ROOT/ (runner:31, check_args_and_update_paths)
2021-11-17 20:41:01,800 [ 1392747 ] INFO : Base configs dir is not set. Will use $SRC_ROOT//programs/server (runner:52, check_args_and_update_paths)
2021-11-17 20:41:01,800 [ 1392747 ] INFO : Cases dir is not set. Will use $SRC_ROOT//tests/integration (runner:59, check_args_and_update_paths)
2021-11-17 20:41:01,800 [ 1392747 ] INFO : src dir is not set. Will use $SRC_ROOT/src (runner:66, check_args_and_update_paths)
2021-11-17 20:41:01,800 [ 1392747 ] INFO : base_configs_dir: $SRC_ROOT/programs/server, binary: $SRC_ROOT/build/programs/clickhouse, cases_dir: $SRC_ROOT/tests/integration  (runner:68, check_args_and_update_paths)
clickhouse_integration_tests_volume
Running pytest container as: 'docker run --net=host -it --rm --name clickhouse_integration_tests --privileged         --volume=$SRC_ROOT/build/programs/clickhouse-odbc-bridge:/clickhouse-odbc-bridge --volume=$SRC_ROOT/build/programs/clickhouse:/clickhouse         --volume=$SRC_ROOT/build/programs/clickhouse-library-bridge:/clickhouse-library-bridge --volume=$SRC_ROOT/build/programs/clickhouse:/clickhouse         --volume=$SRC_ROOT/programs/server:/clickhouse-config --volume=$SRC_ROOT/tests/integration:/ClickHouse/tests/integration         --volume=$SRC_ROOT/src/Server/grpc_protos:/ClickHouse/src/Server/grpc_protos         --volume=/run:/run/host:ro         --volume=clickhouse_integration_tests_volume:/var/lib/docker -e DOCKER_CLIENT_TIMEOUT=300 -e COMPOSE_HTTP_TIMEOUT=600         -e XTABLES_LOCKFILE=/run/host/xtables.lock           -e PYTEST_OPTS=' test_odbc_interaction -ss  -vvv' clickhouse/integration-tests-runner:latest '.
Can't start docker daemon, timeout exceeded.
Traceback (most recent call last):
  File "./runner", line 302, in <module>
    subprocess.check_call(cmd, shell=True)
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'docker run --net=host -it --rm --name clickhouse_integration_tests --privileged         --volume=$SRC_ROOT/build/programs/clickhouse-odbc-bridge:/clickhouse-odbc-bridge --volume=$SRC_ROOT/build/programs/clickhouse:/clickhouse         --volume=$SRC_ROOT/build/programs/clickhouse-library-bridge:/clickhouse-library-bridge --volume=$SRC_ROOT/build/programs/clickhouse:/clickhouse         --volume=$SRC_ROOT/programs/server:/clickhouse-config --volume=$SRC_ROOT/tests/integration:/ClickHouse/tests/integration         --volume=$SRC_ROOT/src/Server/grpc_protos:/ClickHouse/src/Server/grpc_protos         --volume=/run:/run/host:ro         --volume=clickhouse_integration_tests_volume:/var/lib/docker -e DOCKER_CLIENT_TIMEOUT=300 -e COMPOSE_HTTP_TIMEOUT=600         -e XTABLES_LOCKFILE=/run/host/xtables.lock           -e PYTEST_OPTS=' test_odbc_interaction -ss  -vvv' clickhouse/integration-tests-runner:latest ' returned non-zero exit status 1.
```

Reply:
Still the same command works for me.
Try the same docker-compose version as here https://github.com/ClickHouse/ClickHouse/blob/master/docker/test/integration/runner/Dockerfile#L69.

Reply:
I need to restart docker daemon every time after finishing one test run, otherwise it would fail all the time

Reply:
This is strange :(

Reply:
other clues. When start runner, the output from dockerd is
```
failed to start daemon: Error initializing network controller: list bridge addresses failed: PredefinedLocalScopeDefaultNetworks List: [172.16.0.0/24 172.16.1.0/24 172.16.2.0/24 172.16.3.0/24 172.16.4.0/24 172.16.5.0/24 172.16.6.0/24 172.16.7.0/24 172.16.8.0/24 172.16.9.0/24 172.16.10.0/24 172.16.11.0/24 172.16.12.0/24 172.16.13.0/24 172.16.14.0/24 172.16.15.0/24 172.16.16.0/24 172.16.17.0/24 172.16.18.0/24 172.16.19.0/24 172.16.20.0/24 172.16.21.0/24 172.16.22.0/24 172.16.23.0/24 172.16.24.0/24 172.16.25.0/24 172.16.26.0/24 172.16.27.0/24 172.16.28.0/24 172.16.29.0/24 172.16.30.0/24 172.16.31.0/24 172.16.32.0/24 172.16.33.0/24 172.16.34.0/24 172.16.35.0/24 172.16.36.0/24 172.16.37.0/24 172.16.38.0/24 172.16.39.0/24 172.16.40.0/24 172.16.41.0/24 172.16.42.0/24 172.16.43.0/24 172.16.44.0/24 172.16.45.0/24 172.16.46.0/24 172.16.47.0/24 172.16.48.0/24 172.16.49.0/24 172.16.50.0/24 172.16.51.0/24 172.16.52.0/24 172.16.53.0/24 172.16.54.0/24 172.16.55.0/24 172.16.56.0/24 172.16.57.0/24 172.16.58.0/24 172.16.59.0/24 172.16.60.0/24 172.16.61.0/24 172.16.62.0/24 172.16.63.0/24 172.16.64.0/24 172.16.65.0/24 172.16.66.0/24 172.16.67.0/24 172.16.68.0/24 172.16.69.0/24 172.16.70.0/24 172.16.71.0/24 172.16.72.0/24 172.16.73.0/24 172.16.74.0/24 172.16.75.0/24 172.16.76.0/24 172.16.77.0/24 172.16.78.0/24 172.16.79.0/24 172.16.80.0/24 172.16.81.0/24 172.16.82.0/24 172.16.83.0/24 172.16.84.0/24 172.16.85.0/24 172.16.86.0/24 172.16.87.0/24 172.16.88.0/24 172.16.89.0/24 172.16.90.0/24 172.16.91.0/24 172.16.92.0/24 172.16.93.0/24 172.16.94.0/24 172.16.95.0/24 172.16.96.0/24 172.16.97.0/24 172.16.98.0/24
```

The network docker0 is
```
docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default
    link/ether 02:42:91:e3:11:d0 brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.1/20 brd 192.168.15.255 scope global docker0
```

everty time I run the runner, docker0 is deleted.

Reply:
> other clues. When start runner, the output from dockerd is
> 
> ```
> failed to start daemon: Error initializing network controller: list bridge addresses failed: PredefinedLocalScopeDefaultNetworks List: [172.16.0.0/24 172.16.1.0/24 172.16.2.0/24 172.16.3.0/24 172.16.4.0/24 172.16.5.0/24 172.16.6.0/24 172.16.7.0/24 172.16.8.0/24 172.16.9.0/24 172.16.10.0/24 172.16.11.0/24 172.16.12.0/24 172.16.13.0/24 172.16.14.0/24 172.16.15.0/24 172.16.16.0/24 172.16.17.0/24 172.16.18.0/24 172.16.19.0/24 172.16.20.0/24 172.16.21.0/24 172.16.22.0/24 172.16.23.0/24 172.16.24.0/24 172.16.25.0/24 172.16.26.0/24 172.16.27.0/24 172.16.28.0/24 172.16.29.0/24 172.16.30.0/24 172.16.31.0/24 172.16.32.0/24 172.16.33.0/24 172.16.34.0/24 172.16.35.0/24 172.16.36.0/24 172.16.37.0/24 172.16.38.0/24 172.16.39.0/24 172.16.40.0/24 172.16.41.0/24 172.16.42.0/24 172.16.43.0/24 172.16.44.0/24 172.16.45.0/24 172.16.46.0/24 172.16.47.0/24 172.16.48.0/24 172.16.49.0/24 172.16.50.0/24 172.16.51.0/24 172.16.52.0/24 172.16.53.0/24 172.16.54.0/24 172.16.55.0/24 172.16.56.0/24 172.16.57.0/24 172.16.58.0/24 172.16.59.0/24 172.16.60.0/24 172.16.61.0/24 172.16.62.0/24 172.16.63.0/24 172.16.64.0/24 172.16.65.0/24 172.16.66.0/24 172.16.67.0/24 172.16.68.0/24 172.16.69.0/24 172.16.70.0/24 172.16.71.0/24 172.16.72.0/24 172.16.73.0/24 172.16.74.0/24 172.16.75.0/24 172.16.76.0/24 172.16.77.0/24 172.16.78.0/24 172.16.79.0/24 172.16.80.0/24 172.16.81.0/24 172.16.82.0/24 172.16.83.0/24 172.16.84.0/24 172.16.85.0/24 172.16.86.0/24 172.16.87.0/24 172.16.88.0/24 172.16.89.0/24 172.16.90.0/24 172.16.91.0/24 172.16.92.0/24 172.16.93.0/24 172.16.94.0/24 172.16.95.0/24 172.16.96.0/24 172.16.97.0/24 172.16.98.0/24
> ```
> 
> The network docker0 is
> 
> ```
> docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default
>     link/ether 02:42:91:e3:11:d0 brd ff:ff:ff:ff:ff:ff
>     inet 192.168.0.1/20 brd 192.168.15.255 scope global docker0
> ```
> 
> everty time I run the runner, docker0 is deleted.

I found the solution. modify the entrypoint file `docker/test/integration/runner/dockerd-entrypoint.sh`.change the line
```
dockerd --host=unix:///var/run/docker.sock --host=tcp://0.0.0.0:2375 --default-address-pool base=172.0.0.1/12,size=24 &>/ClickHouse/tests/integration/dockerd.log &
```
to 
```
dockerd --host=unix:///var/run/docker.sock --host=tcp://0.0.0.0:2375 --default-address-pool base=192.168.0.1/12,size=24 &>/ClickHouse/tests/integration/dockerd.log &
```